### ISSUE TITLE
modernize packaging

### ISSUE THREAD

**Issue by sneakers-the-rat** [dattalab/moseq2-app#278]
## Is your feature request related to a problem? Please describe.

saw someone recommend this tool, and some labmates were thinking of using it. noticed that install is only possible with docker or conda, and went to see why. the package and its deps aren't terribly complex, but it seems like there is just a lot of tech debt spread out across a number of packages that make it so eg. this package isn't on pypi and isn't as simple to use as `pip install moseq2-app && moseq2` 

other problems:
- looks like a lot of places in the docs have to instruct users to ignore warnings, and resolving those warnings requires updating package versions and code, but that's not really possible to do since things aren't packaged tidily, and there isn't a CD action to make it easy to update for everyone if it did happen. this tech debt only gets worse over time :(
- the versions that are pinned in this project and its dependencies are very old, eg. [numpy 1.18.3](https://github.com/dattalab/moseq2-app/blob/9861ac9803d183537ffcf261f87ba54b480f39bd/setup.py#L37) was released more than 4 years ago. [python 3.7](https://github.com/dattalab/moseq2-app/blob/9861ac9803d183537ffcf261f87ba54b480f39bd/setup.py#L48) reached EOL status more than a year ago, and is no longer receiving security patches. outdated software, particularly high use software like numpy and python, is unsafe to use and should be updated frequently.
- `setup.py` based installations are slow and dangerous - to even check the package metadata, the setup.py file has to be run, so the python world is moving away from them. eg. [this package downloads and extracts a tarball to a `deps` directory under the working directory](https://github.com/sneakers-the-rat/pyhsmm-autoregressive/blob/2a4c73c08dcda959b9bac2f03a2b976dabbc37af/setup.py#L16) when being installed - so if a secondary package were to try and depend on it, any time it was installed that directory would be created no matter where the command was run! thankfully there is no reason to use `setup.py` anymore in a post PEP 517/518/621 world :)
- it's effectively impossible to develop downstream tools when the only means of use are docker containers/conda forge packages or installation has to happen from the git repository - i can't add this as a dependency in my project and use components from it, which enriches the scientific software ecosystem as a whole (as well as extends the use and usefulness of your work!) unless i too make a docker image. the exact version pins make this problem worse because then every downstream package has to have those same outdated version pins.
- it's harder for y'all to develop the tool and make regular releases because the tech debt adds up and each change becomes a harder fight against a sea of outdated dependencies. it might be nice to use some feature from a newer package, but that can't happen if it requires a version of numpy > 1.18.3. I see changes being made in the repo as recently as 3 months ago, but the last released versino was april 2023, same with the docker image.
- it ends up making labor get duplicated, like i see [keypoint-moseq](https://github.com/dattalab/keypoint-moseq) doesn't depend on or use code from this set of packages, so while all of them seem like great tools, they don't build cumulatively and stay maintained, and so a lot of work has to be done twice and labs that use the tools get stuck in a lurch!

## fix

it's not that complicated, but since there is a relatively long chain of deps here that need to be fixed, and some look like they come from outside of the lab and may be abandoned, it requires code to get touched in a bunch of places. 

Here i've updated all the relevant packages to modern packaging tools so that they can all go on pypi and be reused and maintained :)

- https://github.com/sneakers-the-rat/moseq2-app/tree/packaging
- https://github.com/sneakers-the-rat/moseq2-extract/tree/packaging
- https://github.com/sneakers-the-rat/moseq2-model/tree/packaging
- https://github.com/sneakers-the-rat/moseq2-pca/tree/packaging
- https://github.com/sneakers-the-rat/moseq2-viz/tree/packaging
- https://github.com/sneakers-the-rat/pybasicbayes/tree/packaging
- https://github.com/sneakers-the-rat/pyhsmm/tree/packaging
- https://github.com/sneakers-the-rat/pyhsmm-autoregressive/tree/packaging

all of these correctly install (and compile, if they use c extensions), and i got to the point of starting to update calls from outdated packages but stopped to raise this issue and let you know that they are in a state where they can be worked on if you want the changes. i can PR all of these if you'd like. For the packages that aren't from the lab, i could be wrong but most of them seem like they're MIT licensed, so we can just make secondary updated pypi packages for them since they haven't been updated in 4-5 years anyway. 

The tests currently don't pass because of those outdated call issues, but there was a module import in the tests of a module that doesn't seem to exist anymore, so pytest couldn't even start collection, so it seems like the tests maybe don't get run all that often. we can fix that too and ensure that all code always gets tested before committed to default branches :).

we made some nice guides over at pyopensci - https://www.pyopensci.org/python-package-guide/ - and i would be happy to help with the transition since this seems like a relatively common tool in neuro, and software maintenance is a big, hidden part of what makes the work possible. thanks to the group for writing these tools! <3

**Comment by Ctrl-Alt-Karma** [comment]
@sneakers-the-rat 

Just wanted to thank you for doing this and documenting it so thoroughly. I went through the full eight-repository chain you linked, and you were remarkably prescient about the real problem: this was never just an outdated setup.py in moseq2-app, but distributed maintenance debt across all five MoSeq2 packages plus pybasicbayes, pyhsmm, and the autoregressive extension. We are currently qualifying a repaired legacy depth-MoSeq2 pipeline for a large existing study, so we cannot modernize the production environment mid-study; instead, we have frozen the exact Python 3.7/NumPy 1.18 stack, locked the source and dependency identities, repaired correctness and provenance issues, restored the historical test fixtures, and validated the candidate code before a bounded real-recording pilot. During that work we independently ran into many of the conditions you described: floating Git-installed dependencies, fragile cross-repository coupling, stale or non-running tests, legacy API calls, installation side effects, and the crucial fact that “builds successfully” and “scientifically behaves correctly” are separate milestones. Your packaging branches attacked the entire dependency chain rather than papering over the top-level package, and they look like a genuinely valuable foundation for any future modernized MoSeq2 release. It is wild that you assembled all of that, offered to help finish it, and seemingly got no response. Thank you for doing a huge amount of usually invisible scientific-software maintenance work. You were an absolute trooper here.

**Comment by sneakers-the-rat** [comment]
Thanks for saying so. It turns out there is a yawning disconnect between "publishing nature papers about software" and "well developed open source projects you can actually use" and I mostly just try and leave breadcrumbs to lead the former to the latter. Its also extremely common for labs who publish prestige papers about things labeled open source to just never respond to any issue or accept 3rd party contribs, so no offense taken. Not unique to this project, things are like that all over - acknowledging that is relatively rare though, so I appreciate that.

Handling reproducible versions that are tied to specific experiments is much easier when the software is properly packaged and released. Rather than linking to a git repo as is most common in methods sections, it is possible to point to a version or a lockfile with the whole tree frozen. Being able to declare supported ranges of versions is what makes both up- and downstream development possible too: so rather than needing to, say, make a new repository with "moseq 3" in it and do a ton of redundant Dev work either porting things over (Or, more commonly in scientific software, just dropping the whole API), you can just gradually work on it over time while allowing researchers to pin the versions they rely on.

So basically great, glad to see things are moving, and both can be done without conflict - keep a stable version for an existing study while modernizing the packaging so you get that stable surface for free in the future. I'm sure that things have shifted since I sketched this in 2024, probably would have saved some time by merging them then, but we live and we learn. Feel free to cherrypick from these forked repos - would be great to get some credit by preserving git history, but I'm not too precious about that. The goal is the same as it always was: make scientific software run better, so if I can help out just lmk.

**Comment by Ctrl-Alt-Karma** [comment]
@sneakers-the-rat

The version-pinning point is the one that hit hardest, and I want to give you a concrete example of what its absence costs downstream.

`moseq2-model` depends on `pyhsmm`, `pybasicbayes`, and `pyhsmm-autoregressive`, all three pinned to `@master`. That means the ARHMM sampler, which is where essentially all the model math actually lives, is whatever those repos happened to be on the day the environment was built. `moseq2-model` itself is a fairly thin orchestration layer over them.

So the environment that produced a set of results is not reconstructible from its own metadata. We only recovered the exact commits because pip happened to record them in `direct_url.json` under PEP 610:

```
pyhsmm          4e739166746f92bfc968d281f2c1d31e3471409f
pybasicbayes    61f65ad6c781288605ec5f7347efcc5dbd73c4fc
autoregressive  2a4c73c08dcda959b9bac2f03a2b976dabbc37af
```

That worked, but it was archaeology on an installed artifact rather than anything the project told us. If those three were released on PyPI, this is a one-line pin instead of a research project, and the "point at a lockfile in your methods section" thing you described becomes actually possible. Those commits are yours to use if they're a useful starting point for the packaging branches.

You also called something else correctly. You noted pytest couldn't even start collection because of an import of a module that no longer exists. We hit the same wall from the other direction: `skimage.external.tifffile` was removed from scikit-image years ago, and it's still imported. When we ran the historical suites we found a chunk of tests blocked on missing fixtures and a couple more failing on current `cytoolz` behavior. Tests that can't run don't catch regressions, and a project can drift a long way in that state without anyone noticing.

On your last point, I agree completely and it's more or less exactly the plan. We're holding the frozen 3.7/1.18 stack as a reference oracle for the current study rather than upgrading in place, then doing a staged differential comparison against a modern stack with per-stage acceptance criteria. Proper packaging makes that second half dramatically cheaper, because right now "modern stack" means simultaneously replacing NumPy, SciPy, OpenCV, scikit-image, h5py, and pandas and recompiling three unpinned Cython extensions, all in one step. There is no intermediate version where the current pins and anything modern coexist, which turns what should be a sequence of one-variable changes into a single cliff.

Worth flagging separately: alongside the packaging debt we've found correctness issues in the current release branches, dimensional and geometric, that are independent of how the code is installed. Those don't belong in a packaging thread and I don't want to derail yours, so I'll file them properly with reproductions once our validation is finished. Mentioning it only because "it builds" and "the numbers mean what the labels say" turned out to be even further apart than either of us was assuming.

Thanks for the offer to help. I'd like to take you up on it properly, and not only for this.

Once we're clear of the current study's critical path, I'd rather actually work through the packaging transition with you than just cherry-pick from the branches. Selfishly I'd also like to pick your brain more broadly. You've clearly thought a lot harder than I have about what keeps scientific software alive after the paper ships, and I've been learning it the expensive way. If there are other behavioral projects in neuro stuck in this same lurch, I'd be interested in those too.

If there's a better channel than a GitHub thread for that, name it and I'll meet you there. The breadcrumb metaphor is apt, and for what it's worth, this one got picked up.

**Comment by calebweinreb** [comment]
Hello @sneakers-the-rat and @Ctrl-Alt-Karma! Sorry for the radio silence from the Datta lab - the maintainer for this repo left the lab a while ago so monitoring and responding to issues fell through the cracks. Thank you both for all the work you've done helping to modernize the installation and ensure that the pipeline is robust and correct. We'll try to incorporate the packaging improvements that you laid out @sneakers-the-rat.

@Ctrl-Alt-Karma, we definitely want to make sure there are no correctness issues and that the numbers mean what the labels say. Please do share those when you get a chance and let me know if we can help in anyway.

**Comment by Ctrl-Alt-Karma** [comment]
@calebweinreb Thanks for jumping in. The maintainer transition explains the gap.

Here's one finding I can share now, because it's verifiable from source alone and doesn't depend on our pilot. In dattalab/moseq2-extract at 39a6f0f, moseq2_extract/extract/proc.py lines 562 to 564:

python
features['width_mm']  = features['width_px']  * px_to_mm[:, 1]
features['length_mm'] = features['length_px'] * px_to_mm[:, 0]
features['area_mm']   = features['area_px']   * px_to_mm.mean(axis=1)

px_to_mm contains per-axis linear conversion factors in mm per pixel. area_px is a count of image pixels, so multiplying it by the mean of the two linear factors does not produce physical area in mm². The area conversion requires the product of the two per-axis factors.

I'm treating area_mm as intended to represent physical area, and this line as the defect. Worth noting that the units string in util.py reads "Mouse area (mm)" rather than mm², so the implementation and its documented units are both dimensionally inconsistent. If the field was historically intended to represent something other than physical area, that changes the appropriate repair, and I'd rather hear that from the lab than infer it.

We have additional findings across moseq2-extract, moseq2-viz, and moseq2-app. Some are similarly checkable from source. Others need the real-recording pilot before I can responsibly characterize magnitude or downstream effect, and I don't want to put numbers in front of you I can't yet stand behind. I'll keep that inventory out of this packaging thread.

Would you prefer separate issues in each affected repository, or one tracking issue linking out to them?

A candidate correction is already committed on our public moseq2-extract fork at 424d643, but nothing has been proposed against or merged into the Datta lab repositories. I'll link the exact commit with each focused issue rather than pointing you at partially related draft PRs here.

Two things that would help whenever convenient. Archived reference outputs, known-good fixtures, or unpublished definitions behind these calculations, which is what separates deliberate convention from unintended behavior. And a recording containing an object of known dimensions at known depth, if one exists, since that would let the geometry be checked empirically rather than only from the source.

**Comment by Ctrl-Alt-Karma** [comment]
@calebweinreb Filed the first focused correctness issue in `moseq2-extract`:

https://github.com/dattalab/moseq2-extract/issues/179

It includes the source-level dimensional analysis, an executable reproduction against unmodified upstream, the isolated correction, and regression coverage. Same format for the remaining findings as they clear validation.

cc @sneakers-the-rat for continuity with the packaging and test-collection discussion in this thread.

**Comment by Ctrl-Alt-Karma** [comment]
@calebweinreb One other practical follow-up while I have you: we’re nearing the end of recording for the current study and expect to begin processing the remainder soon, so I’m trying to close any external validation questions before we freeze the pipeline we’ll actually run.

Two questions from my earlier comment would be especially helpful:

1. For the remaining correctness findings, would you prefer separate focused issues in the affected repositories, or one tracking issue that links them together?

2. Does the lab have any archived reference outputs, known-good fixtures, calibration recordings, or recordings containing an object of known dimensions/depth that could be shared for validation? Even an old internal reference session would be useful. We can prove several of the defects from source and invariants, but a Datta-origin empirical reference would give us a much stronger check on the geometric corrections before we process the new cohort.

No urgency on packaging from our side right now. I’m specifically trying to eliminate uncertainty around the scientific processing path before the new recordings hit it.

**Comment by calebweinreb** [comment]
1) Probably separate focused issues and maybe you could just add a comment here linking to them?
2) I'm not aware of that type of thing. Probably the closest would be the bucket itself, since we always use the same bucket with the same dimensions. @JRJacoby might know more, although he's on vacation for a few weeks so I can handle any other urgent issues as you nail down the pipeline. Happy to chat on the phone too (617-974-7470) - just let me know if you're planning to call.

**Comment by Ctrl-Alt-Karma** [comment]
Thanks, @calebweinreb - I really appreciate it.

I’ll likely give you a call sometime next week once I get through a few things on my end and have a little cleaner picture of the remaining questions. I’m working through the processing/model-validation side of this now, so I’d rather get those thoughts organized before I waste your time with a half-formed list.

I also have the great privilege of a root canal this Friday, so I’m writing off the end of this week as scientifically unproductive. 😂

I’ll give you a heads-up before I call. Thanks again for being willing to talk through it.
